# pcsct
Plant Single-Cell Comparative Transcriptomics project

To better understand and predict how important energy crops will perform under increasingly erratic environmental conditions, it is necessary to determine how the underlying genes of these organisms behave and contribute to stress responsiveness. For many decades, a great deal of progress in model organisms has enabled some understanding of the genetic machinery that controls stress responses and resistance, yet whether this knowledge translates directly to crops is an open question. With increasing availability of datasets describing transcript abundances in single cells (single-cell datasets), it is now becoming possible to describe more complex relationships between genes of related, but distinct plant species, offering the ability to predict gene function in different crops. For example a gene in the bioenergy grass, Sorghum bicolor (sorghum) might have multiple homologous relatives in Arabidopsis thaliana (a model plant) or Zea mays (maize), yet these relationships are not necessarily functional (i.e. each gene in maize may perform distinct functions or act with a different spatiotemporal pattern than those in Arabidopsis or sorghum). Single-cell transcriptomics data may help to resolve some of these relationships by allowing us to investigate the expression profile of related genes in each species, with the assumption that genes with high sequence similarity that are also expressed in similar tissue/cell types or developmental stages are more likely to be close functional orthologs than genes that do not. This concept of expressologs has been theorized for whole-tissue or whole-organism expression studies, yet has not yet been tested using the much higher resolution single-cell datasets. This project will explore how to integrate single-cell datasets from multiple species together, and use these integrations to investigate the degree to which sequence homology dictates conservation of cell type-specific expression.
